### PR TITLE
[FEATURE] Add Magnetometer measurement to IMU Sensor.

### PR DESCRIPTION
**Pull request by sunkmechie**
## Description
This PR integrates Magnetometer sensor  directly into the existing IMU sensor, upgrading it to a full 9-axis MARG (Magnetic, Angular Rate, and Gravity) sensor.

## Related Issue

Resolves https://github.com/Genesis-Embodied-AI/Genesis/issues/2264

## Motivation and Context

Magnetometers are essential for many robotic applications, especially for orientation estimation and navigation. Adding this completes the standard "9-axis IMU" suite within Genesis, allowing users to simulate more realistic sensor packages for aerial and mobile robots

## How Has This Been / Can This Be Tested?

I have added a new test case in tests/test_sensors.py which:

1. Verifies magnitude invariance (0.5) across different link rotations (Upright, 90° Pitch, and Random Euler angles).
2. Confirms correct axis-swapping logic (e.g., verifying that a 90° pitch correctly transforms the World +Z field into a Local −X reading).
3. Tests both single-environment (n_envs=0) and batched-environment (n_envs=2) scenarios.

Also, I created standalone two scripts ( ```check_magnetometer.py``` and ```visuals_magnetometer.py```)to verify the underlying math and visualization,

```check_magnetometer.py ```
```
import genesis as gs
import numpy as np
import torch

gs.init(backend=gs.cuda)

def run_test(euler_angles, test_name):

    scene = gs.Scene(show_viewer=False)

    plane = scene.add_entity(morph=gs.morphs.Plane())

    cube = scene.add_entity(
        morph=gs.morphs.Box(
            size=(0.2, 0.2, 0.2),
            pos=(0, 0, 1.0),
            fixed=False,
            euler=euler_angles,
        )
    )

    world_field = (0.3, 0.0, 0.4)

    # Add Magnetometer with custom field and slight noise
    imu_sensor = scene.add_sensor(
        gs.sensors.IMU(
            entity_idx=1,
            euler_offset=(0, 0, 0),        
            link_idx_local=0,
            mag_noise=(0.0, 0.0, 0.0),       
            magnetic_field=world_field,
            draw_debug=True,
        )
    )

    scene.build()

    print(f"\n--- Running Test: {test_name} ---")
    print(f"Initial Euler: {euler_angles}")
    print(f"Target World Field: {world_field} (Magnitude: 0.5)")

    # Step once to settle data
    scene.step()

    # Read data
    data = imu_sensor.read()
    # Access the magnetic field (mag) from the 9-DoF IMU data
    mag_tensor = data.mag
    field_np = mag_tensor.detach().cpu().numpy().flatten()

    # Calculate Magnitude: sqrt(x^2 + y^2 + z^2)
    magnitude = np.linalg.norm(field_np)

    print(f"Measured Local Field: x={field_np[0]:.3f}, y={field_np[1]:.3f}, z={field_np[2]:.3f}")
    print(f"Calculated Magnitude: {magnitude:.4f}")

    if np.isclose(magnitude, 0.5, atol=0.01):
        print("VERIFICATION PASSED: Magnitude is invariant.")
    else:
        print("VERIFICATION FAILED: Magnitude mismatch.")

    for _ in range(100):
        scene.step()

    del scene
    
if __name__ == "__main__":
    # Test 1: Upright
    run_test((0, 0, 0), "Upright Position")

    # Test 2: 90 deg pitch (Should move Z field into X or Y channel)
    run_test((0, 90, 0), "90-Degree Pitch")

    # Test 3: Random Orientation
    random_rot = np.random.rand(3) * 360
    run_test(random_rot, "Random Rotation")

```

```visuals_magnetometer.py```

```
import genesis as gs
import numpy as np
import torch


gs.init()


scene = gs.Scene(
    show_viewer=True, 
    viewer_options=gs.options.ViewerOptions(max_FPS=60),
)


plane = scene.add_entity(
    morph=gs.morphs.Plane(),
)

# Add a cube that will fall
cube = scene.add_entity(
    morph=gs.morphs.Box(
        size=(0.2, 0.2, 0.2),
        pos=(0, 0, 1.0), 
        fixed=False,       
        euler=(0, 90, 0)    
    )
)


imu_sensor = scene.add_sensor(
    gs.sensors.IMU(
        entity_idx=1,
        link_idx_local=0,
        mag_noise=(0.01, 0.01, 0.01),
        jitter=0.05,
        draw_debug=True,
    )
)

scene.build()
print("Starting Magnetometer Test...")
print("--------------------------------")
print("Default World Magnetic Field is approx (0, 0, 0.5)")
print("The sensor returns this field rotated into the CUBE's local frame.")
print("Watch the console values change as the cube tumbles!\n")

# Simulation Loop
steps = 200
for i in range(steps):
    scene.step()
    
    data = imu_sensor.read()
    
    if i % 20 == 0:
        field = data.mag 
        print(f"Step {i:03d} | Local Magnetic Field: x={field[0]:.3f}, y={field[1]:.3f}, z={field[2]:.3f}")

print("\nTest Completed Successfully!")

```

## Screenshots (if appropriate):
```check_magnetometer.py```
Screenshot from 2026-01-20 23-12-56

```visuals_magnetometer.py```
Screenshot from 2026-01-20 23-25-57

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by duburcqa** [comment]
I’m wondering whether the magnetometer should be part of the IMU sensor or a dedicated sensor as you did. I tend to frying it should be integrated with IMU.

File: tests/test_sensors.py [hunk 1]
```diff
@@ -474,3 +474,49 @@ def test_raycaster_hits(show_viewer, n_envs):
...
+            magnetic_field=world_field,
+        )
+    )
+
+    scene.build(n_envs=n_envs)
+    scene.step()
+
+    # mag.read() returns a MagnetometerData object
+    data = mag.read()
+    measured_field = data.magnetic_field if hasattr(data, "magnetic_field") else data
```
Comment (YilingQiao) [inline tests/test_sensors.py]: why do we need hasattr here?

File: tests/test_sensors.py [hunk 2]
```diff
@@ -474,3 +474,49 @@ def test_raycaster_hits(show_viewer, n_envs):
...
+    measured_field = measured_field.to(torch.float64)
+
+    magnitudes = torch.norm(measured_field, dim=-1)
+    assert torch.allclose(magnitudes, torch.tensor(0.5, dtype=torch.float64), atol=1e-6)
```
Comment (YilingQiao) [inline tests/test_sensors.py]: use our own `assert_allclose`

File: tests/test_sensors.py [hunk 3]
```diff
@@ -474,3 +474,49 @@ def test_raycaster_hits(show_viewer, n_envs):
...
+
+    # Check directional components (Signs from your successful log)
+    # Expected: x ≈ -0.4, y ≈ 0, z ≈ 0.3
+    val = measured_field[0] if n_envs > 0 else measured_field
+
+    assert torch.abs(val[0] - (-0.4)) < 1e-2
+    assert torch.abs(val[1] - 0.0) < 1e-2
+    assert torch.abs(val[2] - 0.3) < 1e-2
```
Comment (YilingQiao) [inline tests/test_sensors.py]: why only 1e-2 precission?

File: tests/test_sensors.py (hunk 1 above)
Comment (sunkmechie) [reply tests/test_sensors.py]: I apologise, that was leftover code during debug. I now have rewritten the entire test as I integrated the magnetometer into IMU, instead of having it as standalone sensor.

Fixed in 0573b54

File: tests/test_sensors.py (hunk 2 above)
Comment (sunkmechie) [reply tests/test_sensors.py]: Thank you, I have replaced my assert torch.allclose() implementatations with assert_allclose().

Fixed in 0573b54

File: tests/test_sensors.py (hunk 3 above)
Comment (sunkmechie) [reply tests/test_sensors.py]: Updated the tests. now follows the original logic of ```test_imu_sensor()``` 
Fixed in 0573b54

**Comment by sunkmechie** [comment]
> I’m wondering whether the magnetometer should be part of the IMU sensor or a dedicated sensor as you did. I tend to frying it should be integrated with IMU.

Thank you. I’ve updated the implementation to integrate the Magnetometer directly into the IMU class as suggested
0573b54

File: genesis/engine/sensors/imu.py [hunk 4]
```diff
@@ -132,17 +151,31 @@ def build(self):
                 [
                     _get_cross_axis_coupling_to_alignment_matrix(self._options.acc_cross_axis_coupling),
                     _get_cross_axis_coupling_to_alignment_matrix(self._options.gyro_cross_axis_coupling),
+                    _get_cross_axis_coupling_to_alignment_matrix(self._options.mag_cross_axis_coupling),
                 ],
-            ),
-            expand=(self._manager._sim._B, 2, 3, 3),
+            ).unsqueeze(0),
+            expand=(self._manager._sim._B, 1, 3, 3, 3),  # 3 sub-matrices after adding mag
```
Comment (duburcqa) [inline genesis/engine/sensors/imu.py]: I don't want to risk introducing a bug. Are you sure that removing `.unsqueeze(0)` and adding one extra dim is equivalent?

File: genesis/engine/sensors/imu.py [hunk 5]
```diff
@@ -132,17 +151,31 @@ def build(self):
...
+        default_field = self._options.magnetic_field if self._options.magnetic_field is not None else (0.0, 0.0, 0.5)
+        if not isinstance(default_field, torch.Tensor):
+            default_field = torch.tensor(default_field, device=gs.device, dtype=gs.tc_float)
+
+        self._shared_metadata.magnetic_field_vector = concat_with_tensor(
+            self._shared_metadata.magnetic_field_vector,
+            default_field,
+            expand=(self._manager._sim._B, 3),
+            dim=0,
+        )
```
Comment (duburcqa) [inline genesis/engine/sensors/imu.py:171]: Keep all args in one line if it fits.

File: genesis/engine/sensors/imu.py [hunk 6]
```diff
@@ -182,11 +216,16 @@ def _update_shared_ground_truth_cache(
         local_acc = inv_transform_by_quat(acc - gravity[..., None, :], offset_quats)
         local_ang = inv_transform_by_quat(ang, offset_quats)
 
+        n_envs = quats.shape[0]
+        B_world = shared_metadata.magnetic_field_vector.reshape(n_envs, -1, 3)
```
Comment (duburcqa) [inline genesis/engine/sensors/imu.py]: What reshape is supposed to be necessary here? What is the original shape?

File: genesis/engine/sensors/imu.py [hunk 7]
```diff
@@ -208,12 +247,14 @@ def _update_shared_cache(
             shared_metadata.cur_jitter_ts,
             shared_metadata.interpolate,
         )
-        # apply rotation matrix to the shared cache
-        shared_cache_xyz_view = shared_cache.view(shared_cache.shape[0], -1, 3)
-        shared_cache_xyz_view.copy_(
-            torch.matmul(shared_metadata.alignment_rot_matrix, shared_cache_xyz_view.unsqueeze(-1)).squeeze(-1)
-        )
-        # apply additive noise and bias to the shared cache
```
Comment (duburcqa) [inline genesis/engine/sensors/imu.py:216]: Why did you have to change this?

File: tests/test_sensors.py [hunk 8]
```diff
@@ -102,11 +112,18 @@ def test_imu_sensor(show_viewer, tol, n_envs):
     with np.testing.assert_raises(AssertionError):
         assert_allclose(lin_acc_no_offset, lin_acc_with_offset, atol=0.2)
     imu.set_pos_offset((0.0, 0.0, 0.0))
+    # check that entity rotation affects magnetometer readings
```
Comment (duburcqa) [inline tests/test_sensors.py]: Skip line between check blocks.

File: tests/test_sensors.py [hunk 9]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
     box.set_COM_shift([0.0, 0.0, 0.0])
-    box.set_quat([0.0, 0.0, 0.0, 1.0])
+    box.set_quat(torch.tile(torch.tensor([1.0, 0.0, 0.0, 0.0]), (*batch_shape, 1)))
 
     # wait for the box to be stationary on ground
     for _ in range(50):
         scene.step()
 
-    assert_allclose(imu.read().lin_acc, (0.0, 0.0, -GRAVITY), tol=5e-6)
+    assert_allclose(imu.read().lin_acc, torch.tensor([0.0, 0.0, -GRAVITY]), tol=5e-6)
```
Comment (duburcqa) [inline tests/test_sensors.py]: Remove this.

File: tests/test_sensors.py [hunk 10]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
     assert_allclose(imu.read().ang_vel, (0.0, 0.0, 0.0), tol=1e-5)
 
-    # rotate IMU 90 deg around x axis means gravity should be along -y axis
+    # rotate IMU 90 deg around x axis means gravity should be along -y axis, mag in sensor frame: (0.3, 0.4, 0.0)
     imu.set_quat_offset(gu.euler_to_quat((90.0, 0.0, 0.0)))
-    imu.set_acc_cross_axis_coupling((0.0, 1.0, 0.0))
+    imu.set_acc_cross_axis_coupling((0.0, 0.0, 0.0))
```
Comment (duburcqa) [inline tests/test_sensors.py]: Remove this.

File: tests/test_sensors.py [hunk 11]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
     scene.step()
-    assert_allclose(imu.read().lin_acc, GRAVITY, tol=5e-6)
-    imu.set_quat_offset((0.0, 0.0, 0.0, 1.0))
+    assert_allclose(imu.read().lin_acc, torch.tensor([0.0, -GRAVITY, 0.0]), tol=5e-6)
+    assert_allclose(imu.read().mag, torch.tensor([0.3, 0.4, 0.0]), tol=1e-6)
```
Comment (duburcqa) [inline tests/test_sensors.py]: Remove torch.tensor as just pass type.

File: tests/test_sensors.py [hunk 12]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
     assert_allclose(imu.read().ang_vel, (0.0, 0.0, 0.0), tol=1e-5)
 
-    # rotate IMU 90 deg around x axis means gravity should be along -y axis
+    # rotate IMU 90 deg around x axis means gravity should be along -y axis, mag in sensor frame: (0.3, 0.4, 0.0)
     imu.set_quat_offset(gu.euler_to_quat((90.0, 0.0, 0.0)))
-    imu.set_acc_cross_axis_coupling((0.0, 1.0, 0.0))
+    imu.set_acc_cross_axis_coupling((0.0, 0.0, 0.0))
     scene.step()
-    assert_allclose(imu.read().lin_acc, GRAVITY, tol=5e-6)
-    imu.set_quat_offset((0.0, 0.0, 0.0, 1.0))
```
Comment (duburcqa) [inline tests/test_sensors.py:134]: Do not remove this.

File: tests/test_sensors.py [hunk 13]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
     imu.set_acc_cross_axis_coupling((0.0, 0.0, 0.0))
-
+   
     scene.reset()
-
-    assert_allclose(imu.read().lin_acc, 0.0, tol=gs.EPS)  # biased, but cache hasn't been updated yet
-    assert_allclose(imu_delayed.read().lin_acc, 0.0, tol=gs.EPS)
-    assert_allclose(imu_noisy.read().ang_vel, 0.0, tol=gs.EPS)
-
-    imu.set_bias(BIAS + (0.0, 0.0, 0.0))
```
Comment (duburcqa) [inline tests/test_sensors.py:143]: Remove this changes.

File: tests/test_sensors.py [hunk 14]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
+  
+    zero_param = torch.zeros((*batch_shape, 3))
```
Comment (duburcqa) [inline tests/test_sensors.py]: Don't do this.

File: tests/test_sensors.py [hunk 15]
```diff
@@ -117,32 +134,35 @@ def test_imu_sensor(show_viewer, tol, n_envs):
...
+    assert_allclose(imu.read().lin_acc, zero_param, tol=gs.EPS)
+    assert_allclose(imu_delayed.read().lin_acc, zero_param, tol=gs.EPS)
+    assert_allclose(imu_noisy.read().ang_vel, zero_param, tol=gs.EPS)
+   
+    imu.set_bias(BIAS + (0.0, 0.0, 0.0) + (0.05, 0.05, 0.05))
     scene.step()
     assert_allclose(imu.read().lin_acc, BIAS, tol=tol)
+    assert_allclose(imu.read().mag, torch.as_tensor(MAG_FIELD) + 0.05, tol=tol)
```
Comment (duburcqa) [inline tests/test_sensors.py]: Do not cast to tensor.

File: tests/test_sensors.py
Comment (duburcqa) [inline tests/test_sensors.py:1]: Do not change what the existing unit test is doing, unless you can explain why this is necessary.

You can add additional checks for magnometer but there is no reason to update what is already there.

File: genesis/engine/sensors/imu.py (hunk 4 above)
Comment (sunkmechie) [reply genesis/engine/sensors/imu.py]: My mistake, adding the extra dim was unnecessary, I modified it now to 

```  expand=(self._manager._sim._B, 3, 3, 3), ```
            ```dim=1,```

File: genesis/engine/sensors/imu.py (hunk 6 above)
Comment (sunkmechie) [reply genesis/engine/sensors/imu.py]: refactored the logic to follow acc and gyro, reshape isn't needed anymore

**Comment by sunkmechie** [comment]
@duburcqa Thank you for the corrections. I removed all the unnecessary typecasts and changed the MAG_FIELD implementation so that it doesn't require weird reshape anymore.

I found that the Identity quat was wrong in the original code and assumed [x,y,z,w] instead of the [w,x,y,z]; I fixed it and verified with ```pytest -v --forked tests/test_sensors.py```. 

Please verify if my change was correct. My assumption was that this was previously uncaught as a 180-degree rotation in box doesn't affect gyro (spin vec direction didn't change) and accl readings. But as the mag is sensitive to this, was able to catch the bug as the sign difference was clear now. 

81a1116

**Review by duburcqa — APPROVED** [review]
